### ISSUE TITLE
Provide mapping into java.util.Map using enum values as key

### ISSUE THREAD

**Issue by pe-st** [mapstruct/mapstruct#717]
I'd like to map the fields of a bean class into a dictionary-like class, using [MapStruct][ms]. My source class is a standard bean (simplified example):

    public class Bean {
        private String a;
        private String b;
    
        public String getA() {
            return a;
        }
    
        public void setA(String a) {
            this.a = a;
        }
    
        public String getB() {
            return b;
        }
    
        public void setB(String b) {
            this.b = b;
        }
    }

Now I want to map these fields into a Map-like container:

    public class Dict {
    
        public enum Tag {
            A,
            B
        }
    
        private Map<Tag, String> dict = new HashMap<>();
    
        public String getEntry(Tag tag) {
            return dict.get(tag);
        }
    
        public void setEntry(Tag tag, String s) {
            dict.put(tag, s);
        }
    }

In other words, I'd like [MapStruct][ms] to generate something along the lines of:

        target.setEntry(Dict.Tag.A, source.getA());
        target.setEntry(Dict.Tag.B, source.getB());

I couldn't find anything similar in the MapStruct [documentation][doc]. There is much flexibility for getting at mapping sources (nested sources, expressions), but for targets I can see only the `target = "propertyname"` notation which doesn't leave much room for flexibility.

(original question from [stackoverflow][so], added as issue as requested by Gunnar)

 [ms]: http://mapstruct.org/
 [doc]: http://mapstruct.org/documentation/
 [so]: http://stackoverflow.com/q/34560093/3686

**Comment by steirer00** [comment]
If this feature is added I would suggest that the other way around should also be done. Meaning: Map -> Bean , Set -> Bean

`java.util.Map<String, String> myMap();` with entry {"last_name","some name"}

```
public class Bean{

String firstName;
String lastName;

// getters + setters
}

```
```
@Mapper
public interface Mapper{

@Mapping(source="last_name", target="lastName")
public Bean maptoBean (java.util.Map<String,String> myMap)

}
```
additionally would be very usefull to use for HttpServletRequest e.g. `Map<String, String[]> parameters = request.getParameterMap();` 
or other combinations like `Map<String, List<String>>`, ...

**Comment by tomliuxd** [comment]
I am looking for something similar (if not the same)

1) Mapping bean property to target map with given key name
2) Mapping target map with given key name to bean property

for example
`
abc
def
`
public class ClassA {
    Map<String, String> properties;
}

public class ClassB {
    String serverName;
}

@Mapping(target="properties(key='server name')", source="serverName")
public ClassA b2a (ClassB b);

@Mapping(target="serverName", source="properties(key='server name'")
public ClassB a2b (ClassA a);

**Comment by apolit** [comment]
Similar functionality is available in Dozer: http://dozer.sourceforge.net/documentation/mapbackedproperty.html
It would be great to have it in MapStruct

**Comment by tomliuxd** [comment]
actually we are using dozer now.  but its performance is not good.  we are considering to move to either mapstruct or other tools.

**Comment by pehunka** [comment]
I would like to see this feature in MapStruct as well, if possible.

**Comment by wmelon84** [comment]
People are supposed to be using https://github.com/FasterXML/jackson for that purpose. Never tested it!
It would be great to have MapStruct builtin feature.
